### PR TITLE
Product Collection: Trigger `wc-blocks_product_list_rendered` JS event

### LINKED ISSUE #48862: Product Collection: trigger `wc-blocks_product_list_rendered` JS event
> [!WARNING]  
> The issue has been repurposed. But I'm keeping the original description for context. Read the update below!

## Original issue:

**Issue Description:**
In All Products, the `experimental__woocommerce_blocks-product-list-render` action is fired when All Products or AbstractProductGrid blocks are rendered or re-rendered. Implemented for tracking purposes for extension of [Google Analytics for WooCommerce](https://woocommerce.com/products/woocommerce-google-analytics/) plugin.

PC doesn’t fire this action. We could do it in the same manner as AbstractProductGrid does (check the next section) although it attaches DOM events listeners, which we should be avoiding with the Interactivity API in mind. Worth checking if we could utilize Interactivity API for that (e.g. [wp-on-document-async](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/api-reference/#wp-on-document-async)).

**Goal**
Implement the event in PC block

Note: API is experimental and event payload is not documented. Still, it should be considered a public API and align with the payload provided in other places (All Products, AbstractProductGrid) as extensions rely on it.

Ref: GitHub discussion: https://github.com/woocommerce/woocommerce/discussions/48814

## Updated goal

Based on the discussion: https://github.com/woocommerce/woocommerce/discussions/48814, we decided the users of Product Collection block should rely on JS events:

I'm repurposing this issue and instead of `experimental__woocommerce_blocks-product-list-render`, the JS event `wc-blocks_product_list_rendered` broadcasted in a similar manner as:

- `wc-blocks_adding_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L281-L282
- `wc-blocks_added_to_cart` -> https://github.com/woocommerce/woocommerce/blob/f64f44908edeb0f922c478ad71ae41886f33f84e/plugins/woocommerce-blocks/assets/js/data/cart/actions.ts/#L292-L293

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Emit the JS event when Product Collection is rendered or re-rendered (e.g. when changing a page). Event contains information about collection that triggered it.

Closes https://github.com/woocommerce/woocommerce/issues/48862

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. Use some plugin to include custom JS script on the page e.g. [Easy Code Snippets](https://wordpress.org/plugins/easy-code-snippets/)
2. Attach this code:

```
document.body.addEventListener(
	'wc-blocks_product_list_rendered',
	e => console.log('Event triggered', e)
);
```

If you're using Easy Code Snippets, don't forget to wrap it in `<script></script>` tag and set the position to "Footer":
image

#### Testing
1. Create new post
2. Add Product Collection block
3. Choose any collection
4. Save and go to frontend
5. Open browser's console
6. **Expected:** One PC block is expected to trigger a single event so there should be one "Event triggered" entry in the console with event details. (look at the screenshot)
7. **Expected:** Expand the event and look for `detail.collection`. It should be a string representing collection you inserted: `woocommerce/product-collection/<<COLLECTION>>`:

Screen Shot 2024-08-02 at 16 48 55 PM

9. Go back to Editor and add 2 more Product Collection blocks and choose some collections
10. Save and go to frontend and open browser's console
11. **Expected:** There should be 3 events triggered, one per each PC block.
12. Change a page in one of the collections
13. **Expected:** Another event has been triggered

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: For stability, I'd recommend using `expect.poll` here so that we don't have to use `waitForLoadState` which is not very reliable:

```js
await pageObject.publishAndGoToFrontend();
await expect.poll( async () => await page.evaluate( 'window.eventFired' ) ).toBe( 1 );
```

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: Ditto!

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: I'd use the event ID in the title - easier to find a responsible test! For example `"emits the wc-blocks_product_list_rendered event"`.

**Review by WunderBart — APPROVED** [review]
LGTM! 🚀

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts]: 🔥 😄 

```suggestion

```

File: plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts]: Just out of curiosity, why are we adding this new event in the `legacy-events.ts` file? Is this new event a legacy event, or are we just adding it to this file because it's related to events? 🙂

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]:
```suggestion
	private function enable_client_side_navigation( $block_content ) {
```

File: plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts]: >Is this new event a legacy event, or are we just adding it to this file because it's related to events?

The latter. I don't think creating a new place for it would help in maintenance so wanted to keep them in one place.

However, in the meantime, I kicked off a refactor on that area: https://github.com/woocommerce/woocommerce/pull/50199 (branched out from this branch) to extract a separate `@woocommerce/base-events` package. I'm discussing it with @gigitux elsewhere. But the reactor is not blocking delivery of the event itself hence I'm keeping the separated.

I hope that makes sense

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx]: I am not sure if we should add `collection` attribute in the default block. Till now, I assumed that we won't have a collection attribute in the default block. Because of this, some conditions in the code check if collection exists or not. ([Example](https://github.com/woocommerce/woocommerce/blob/150cf84850d92e6267792fbb21b3cc3c9560596e/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx#L105)). In case, we really wanna have collection attribute in the default block, we should update these conditions too.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: Nit: We are repeating this at 3 places in the code. It would be great if we can somehow make it DRY. 
```PHP
$p->next_tag( array( 'class_name' => 'wp-block-woocommerce-product-collection' ) )
```

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (imanish003) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: Super Nit: I am unsure if `data-wc-context` belongs to `add_init_directive` function. Probably, renaming this function will help make the code more readable 🙂

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/types.ts
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/types.ts:138]: Thanks for removing it 💜

File: plugins/woocommerce-blocks/docs/third-party-developers/extensibility/product-collection/dom-events.md
Comment (imanish003) [inline plugins/woocommerce-blocks/docs/third-party-developers/extensibility/product-collection/dom-events.md:1]: Thanks for adding documentation for the new event 💜 Should we consider adding an example of how to use this event in the documentation?

File: plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts
Comment (imanish003) [inline plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:1612]: TIL That we can add `addInitScript` in the e2e tests. This is super cool!

File: plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/base/utils/legacy-events.ts]: Totally makes sense! Thanks you for kicking off the refactor. I really like how you are creating ENUM for event names 💜

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hey @kmanijak!

I left some comments on the PR. Most of them are nitpicks, except for the comment about adding a collection attribute in the default block. I am happy to approve the PR once we discuss that comment. Good work on this PR! 💜

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/utils.tsx]: You're right! I'm removing and I'll follow the same approach in sending event adding `CoreCollectionNames.PRODUCT_CATALOG` if there's no `collection` attribute! 👍

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: Good point! Updated!

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: I changed the function name to be more specific, good point Manish!

File: plugins/woocommerce-blocks/docs/third-party-developers/extensibility/product-collection/dom-events.md
Comment (kmanijak) [reply plugins/woocommerce-blocks/docs/third-party-developers/extensibility/product-collection/dom-events.md:1]: Agreed! I provided an example and made overall cleanup of this file to be simpler. I also moved that to `docs/` directory to make it public facing in Developer Blog.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: Nitpick: I can see [here](https://github.com/woocommerce/woocommerce/blob/dd34694f9214bf368588fa940e9ff3669a7ecf62/plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx#L193-L195) that we are using `collection: collection || CoreCollectionNames.PRODUCT_CATALOG` i.e.
```tsx
triggerProductListRenderedEvent( {
  collection: collection || CoreCollectionNames.PRODUCT_CATALOG,
} );
```

Shouldn't we do same here too? 
OR Do we even need to pass `PRODUCT_CATALOG`? Maybe we can assume that if collection is not passed, it's a custom collection. With new flow to add Product Collection, there isn't any user facing word for Product Catalog. Maybe slowly we can even deprecate it? My reasoning is that when user clicks on "Create your own", we don't need to call it Product Catalog. What do you think?

File: docs/product-collection-block/dom-events.md
Comment (imanish003) [inline docs/product-collection-block/dom-events.md:29]: Thanks for adding example in the documentation 🙌🏻

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: ### Major - Handle the scenario when Product Collection block isn't visible

I noticed that `wc-blocks_product_list_rendered` event is triggered even if Product Collection block isn't there on a post as you can see in screenshot below. It was surprising therefore I debugged it. I found out that there was "Top Rated" collection in Empty Mini Cart view, which is always rendered on the page. So, the event was triggered.

IMO, it's important to handle this scenario. One solution could be to verify in the `onRender` function whether the Product Collection block is visible in the viewport. If it's not visible, we shouldn't trigger the event.

What do you think? 

image

image

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hey Karol!

I tested the PR and most of the things are working as expected. I have left one major feedback that I think we should address before merging the PR. Apart from that, I have left some nitpicks in the PR, please feel free to address them if you think they are valid. 🙇 

Good work so far! We are almost there 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: > OR Do we even need to pass PRODUCT_CATALOG? Maybe we can assume that if collection is not passed, it's a custom collection. With new flow to add Product Collection, there isn't any user facing word for Product Catalog. Maybe slowly we can even deprecate it? My reasoning is that when user clicks on "Create your own", we don't need to call it Product Catalog. What do you think?

What would you send in such a case, no payload? Or `collection: undefined`?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: > One solution could be to verify in the onRender function whether the Product Collection block is visible in the viewport. If it's not visible, we shouldn't trigger the event.

But that means excluding many blocks from triggering the event. For example, the PC block rendered in the footer is not visible at first sight. And I don't think creating an observer to send an event when the block comes into viewport is the way we'd like to go with 🤔

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: BTW I moved the default to `ProductCollection.php` [here](https://github.com/woocommerce/woocommerce/pull/50166/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R297)

Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: IMO, we can send the collection as `undefined` in such case for consistency. What do you think?

I think we wanna communicate that it's a product collection but not a specific collection.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: Oh, my bad! I used the wrong terminology, i.e., viewport 😅 I meant to say that we should check if the block is visible anywhere on the page, which means it doesn't have to be in the viewport. 

1. We will need to check if element is hidden using CSS i.e.
```tsx
function isVisible(element) {
  const style = window.getComputedStyle(element);
  return style.display !== 'none' && style.visibility !== 'hidden' && style.opacity !== '0';
}
````

2. We will need to check top, left, right, bottom of the element to see if it's visible on the page or not. Something like this:
```tsx
function isInViewport(element) {
  const rect = element.getBoundingClientRect();
  return (
    rect.top <= (window.innerHeight || document.documentElement.clientHeight) &&
    rect.left <= (window.innerWidth || document.documentElement.clientWidth) &&
    rect.bottom >= 0 &&
    rect.right >= 0
  );
}
```

2nd is the case with mini cart block which uses drawer. 

---

I will leave it to you to decide if we should handle this scenario or not. Because if we will decide to handle it then we can't use `data-wc-init` attribute to trigger the event, as it won't be re-triggered when we open the mini cart and the block becomes visible. We will probably need to use `data-wc-run`. However, this may result in more complex code. Therefore, I will leave it to you to decide if we should handle this scenario or not 🙂

Comment (xristos3490) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: Interesting idea, @imanish003! My two cents here would be to leave this as it is and, if necessary, streamline it as you suggested.

I'm saying this because of potential third-party cases where the block might be hidden for some custom reason, and we cannot 100% control this. So, leaving it simple, I think, is best for now.

**Review by xristos3490 — COMMENTED** [review]
Works as expected, @kmanijak!

Out of curiosity, wouldn't it make sense to have the srcElement or the target element as the actual PC DOM element? Is that even possible here? How do you imagine using this without having a reference to the DOM? 🤔

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: I double-checked and previous blocks (All Products, Newest Products etc) trigger action on render no matter whether visible or not (for example if placed in Mini Cart). I'd keep it as-is, same as suggested by @xristos3490. Covering all of the edge cases would make it extremely complex IMO.

**Comment by kmanijak** [comment]
> Out of curiosity, wouldn't it make sense to have the srcElement or the target element as the actual PC DOM element? Is that even possible here? How do you imagine using this without having a reference to the DOM? 🤔

@xristos3490, what would you use the DOM reference for? My understanding is that this event is broadcast to communicate that collection is rendered. Emitting an event from the actual PC element means someone should attach an event listener to the same element, making it totally unusable. Hence we emit this on a `document` level. It's just informative, listener can react to that but I don't think we should provide information about DOM element.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: @imanish003 with that decision, does it mean it could be merged?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: > IMO, we can send the collection as undefined in such case for consistency. What do you think?
 
Do you mean consistency with us not exposing this name externally? Or consistency with what?

Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: Yeah, consistency with us not exposing this name externally. IMO We don't need to expose this name externally. We don't need to call it a Product Catalog collection. Internally, we can call it a Product Collection or custom collection. Does it make sense? Maybe I am overthinking it 😅

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx]: I am good with the decision to keep it as it is 🤝 

I think this can be merged after making a decision [here](https://github.com/woocommerce/woocommerce/pull/50166#discussion_r1714814545) 🙂

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: I like what you're saying and agree we only use this name internally. I'll make an update!

**Review by imanish003 — APPROVED** [review]
Thanks @kmanijak for your patience and hard work on this PR 🙇‍♂️ I think we can merge this PR after making a decision [here](https://github.com/woocommerce/woocommerce/pull/50166#discussion_r1714814545). I am pre-approving the PR. Great work! 🚀

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:145]: Updated in https://github.com/woocommerce/woocommerce/pull/50166/commits/415420b3931bfb35b4f9bc3929ed01bb3f6340a8. Also, updated the docs in https://github.com/woocommerce/woocommerce/pull/50166/commits/ce30c3baf7aa0b0547ccd618e1394f5079359dd3

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (roykho) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:229]: This causes a warning when `data-wc-context` is not found. We need to first check before use. Maybe something like this?
`$current_context = ! empty( $p->get_attribute( 'data-wc-context' ) ) ? json_decode( $p->get_attribute( 'data-wc-context' ), true ) : [];`

Comment (kmanijak) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:229]: Thanks @roykho for catching it! @wavvves already let me know of this and it's been fixed in https://github.com/woocommerce/woocommerce/pull/50661, pull the newest `trunk` and a warning will be gone!

**Pull request merged**
